Render jobs from templates (#1363)
kyma-project/test-infra: docs/prow/templates.md

# Templates

> Jobs and Prow configuration are generated from templates that are in `templates` directory. There are following items there:

Reviewer: ```suggestion
Jobs and Prow configuration are generated from the templates in the `templates` directory. The directory contains these items:
```

> - `config.yaml` - config file that describes files to generate

Reviewer: ```suggestion
- `config.yaml` is the config file that describes files to generate.
```

> - `templates` - directory to gather all template files

Reviewer: ```suggestion
- `templates` is the directory to gather all the template files.
```

> ### Config structure

Reviewer: Should we use the full word instead of `config`?

> Config file is expected to have two keys:

Reviewer: ```suggestion
The config file has two keys:
```
Is expected or has?

Author: Currently it has, but at the same time it has to :D

> - `global` - a map of values available for all templates

Reviewer: ```suggestion
- `global` is a map of values available for all templates.
```

> - `templates` - a list of files to generate

Reviewer: ```suggestion
- `templates` is a list of files to generate.
```

> Complete structure can be found [here](../../development/tools/cmd/rendertemplates/main.go).

Reviewer: ```suggestion
To see the complete structure, go [here](../../development/tools/cmd/rendertemplates/main.go).
```

> Other then that `config.yaml` uses YAML techniques for better organization, so normal YAML rules apply.

Reviewer: ```suggestion
Other then that, `config.yaml` uses YAML techniques for better organization, so normal YAML rules apply.
```
This sounds somewhat too colloquial... 
Maybe change it to something like:
> `config.yaml` uses YAML techniques for better organization. If you use YAML, you are already familiar with the rules. 

?

Author: On the second thought, if you see it is a YAML you may expect all the YAML things to be there, so I removed the sentence completely.

### Template development

> A template will receive two objects as inputs:

Reviewer: ```suggestion
A template receives two objects as input:
```
Don't use future tense (_will_), use present tense where possible.

> - `Values` which contains all `values` specified in its configuration in config file

Reviewer: ```suggestion
- `Values`, which contains all the values specified under `values` in the configuration file
```

> - `Global` which contains values specified under `global` in config file

Reviewer: ```suggestion
- `Global` which contains values specified under `global` in the configuration file.
```

> All functions from [`sprig`](https://github.com/Masterminds/sprig) library are available in templates. It is the same library that is used by helm, so if you know helm there's nothing new. Also some additional functions are available:

Reviewer: ```suggestion
All the functions from [`sprig`](https://github.com/Masterminds/sprig) library are available in the templates. It is the same library that is used by helm, so if you know helm, you are already familiar with them. Also, a few additional functions are available:
```

Also, do we capitalize `helm`?

> - `releaseMatches <release> <since> <until>` - returns a bool value indicating if `release` fits in the range. You can use `nil` to remove one of the bounds. Example: `releaseMatches {{ $rel }} '1.2' '1.5'` will check if release `$rel` is younger then `1.2` and older then `1.5`.

Reviewer: ```suggestion
- `releaseMatches <release> <since> <until>` returns a boolean value indicating whether `release` fits in the range. Use `nil` to remove one of the bounds. For example: `releaseMatches {{ $rel }} '1.2' '1.5'` checks if the release `$rel` is younger than `1.2` and older than `1.5`.
```

> - `matchingReleases <all-releases> <since> <until>` - returns a filtered list of releasesto only those that fits in the range.

Reviewer: ```suggestion
- `matchingReleases <all-releases> <since> <until>` returns a list of releases filtered to only those that fit in the range.
```

> ### Add new component

Reviewer: ```suggestion
### Add a new component
```

> To add new component find a `templates` entry for `templates/component.yaml`. Then add new entry with your component to the `render` list. An example below defines a component in Kyma repository using go1.12 buildpack:

Reviewer: ```suggestion
To add a new component, find a `templates` entry for `templates/component.yaml`. Then, add a new entry with your component to the `render` list.  
This example defines a component in the Kyma repository using go1.12 buildpack:
```

```yaml
  - from: templates/component.yaml
    render:
      - to: ../prow/jobs/kyma/components/new-component/new-component.yaml
        values:
          <<: *go_kyma_component_1_12
          path: components/new-component
    ...
```

> If buildpack you're willing to use is not there yet you have to add it. The best would be to follow existing buildpacks.

Reviewer: ```suggestion
If the buildpack you want to use is not there yet, add it. If possible, use existing buildpacks.
```
- It would be best to inform the user why they should refrain from using other buildpacks. If only to avoid the hassle, consider removing the second sentence. 
- `You're willing to` has a slightly different meaning. See [here](https://dictionary.cambridge.org/dictionary/english/willing).

Author: I'd rather meant to suggest following the style in which other buildpacks are defined. I'll try to rephrase it.

> When writing tests for new component use function `tester.GetKymaReleasesSince(<next release>)` to create release jobs tests. This will automatically check if new release jobs where created when doing release.

Reviewer: ```suggestion
When writing tests for the new component, use the function `tester.GetKymaReleasesSince(<next release>)` to create release jobs tests. This automatically checks whether new release jobs were created when doing release.
```

### Change component job configuration

> Whenever you need to change component job configuration follow this steps:

Reviewer: ```suggestion
To change component job configuration, follow these steps:
```
Add the appropriate article (`a`/`the`) before `component job configuration`.

> 1. In `config.yaml` change the name of the file where the jobs are generated. A suffix like `deprecated` should be enough. Change this path also in tests.

Reviewer: ```suggestion
1. In `config.yaml`, change the name of the file where the jobs are generated. For example, add the suffix `deprecated`. Change this path in tests as well.
```

- Please don't speculate. Don't use `should`. Either it is enough or it's not. 
- Avoid using passive voice when possible. Can you rephrase `file where the jobs are generated` to something like:
`file containing the generated jobs`?

# Templates

Jobs and Prow configuration are generated from templates stored in the `templates` directory. The directory contains these items:

- `config.yaml` is the config file that describes files to generate.
- `templates` is the directory to gather all the template files.

### The configuration file structure

The config file has two keys:

- `global` is a map of values available for all templates.
- `templates` is a list of files to generate.

To see the complete structure, see [this](../../development/tools/cmd/rendertemplates/main.go) file.

### Template development

A template receives two objects as input:

> - `Values`, which contains all the values specified under `values` in the configuration file

Reviewer: ```suggestion
- `Values`, which contains all the values specified under `values` in the configuration file.
```

### Add a new component

To add a new component, find a `templates` entry for `templates/component.yaml`. Then, add a new entry with your component to the `render` list.  
This example defines a component in the Kyma repository using the `go1.12` buildpack:
```yaml
  - from: templates/component.yaml
    render:
      - to: ../prow/jobs/kyma/components/new-component/new-component.yaml
        values:
          <<: *go_kyma_component_1_12
          path: components/new-component
    ...
```

If the buildpack you want to use is not there yet you have to add it. When you add a new buildpack follow the example of already defined ones.

When writing tests for a new component, use the `tester.GetKymaReleasesSince(<next release>)` function to create release jobs tests. This automatically checks whether new release jobs were created in the release process.

### Change component job configuration

To change component job configuration, follow these steps:
1. In the `config.yaml` file, change the name of the file where the jobs are generated. For example, add the suffix `deprecated`. Change the path to this file in tests accordingly.
2. Add `until: <last release>` to this configuration.

> 3. Create a new entry with a new configuration. It should generate new jobs to the file used before.

Reviewer: ```suggestion
3. Create a new entry with the new configuration. Set the `to` field to point to the file responsible for storing jobs.
```

4. Add `since: <next release>` to new entry.

> Example: buildpack for api-controller has changed from go1.11 to go.12 in release 1.5. Before the change component configuration looked like this:

Reviewer 2: ```suggestion
Example: buildpack for the API Controller has changed from `go1.11` to `go.12` in release 1.5. This is the component configuration before the buildpack change:
```

```yaml
      - to: ../prow/jobs/kyma/components/api-controller/api-controller.yaml
        values:
          <<: *go_kyma_component_1_11
          path: components/api-controller
```

> New configuration will look like follows:

Reviewer 2: ```suggestion
This is what the configuration created after the buildpack change looks like:
```
